content/sips/sip-399.md

```diff
@@ -0,0 +1,103 @@
+network: Base
+status: Draft
+author: 'MEB (@barrasso)'
+created: 2024-07-16
+---
+
+## Simple Summary
+
+Deploy a Flash Loan utility contract to enable users to close their Perps V3 positions in a single transaction.
+
+## Abstract
+
+This SIP proposes to deploy a new contract, `PerpsFlashLoanUtil`, which will allow users to close their Perps V3 positions by utilizing flash loans. This contract interacts with the Synthetix proxies and Uniswap's SwapRouter to facilitate the necessary operations. The contract also includes a utility function to set Uniswap pool fees dynamically based on the collateral or margin type.
+
+## Motivation
+
+The motivation behind this proposal comes from requests by Perps integrators and partners who seek a better user experience when closing Perps positions. The current process can be cumbersome, requiring multiple steps and transactions. This flash loan utility contract aims to streamline the process, allowing users to close positions atomically, thereby improving efficiency and user satisfaction.
+
+## Specification
+
+### Overview
+
+The `PerpsFlashLoanUtil` contract will allow users to close their Perps positions using flash loans from the Aave protocol. The process involves wrapping USDC to snxUSD, repaying debt, modifying collateral, unwrapping to the original collateral type, swapping to USDC, and repaying the flash loan.
+
+### Rationale
+
+The proposed solution leverages existing Synthetix and Uniswap contracts to facilitate the flash loan process. By using a flash loan, users can close their Perps positions without needing to front the necessary collateral. The contract allows for dynamic setting of Uniswap pool fees and supporting various collateral types.
+
+### Technical Specification
+
+#### Interfaces
+
+```solidity
+/**
+ * @title Interface for the PerpsFlashLoanUtil contract.
+ * @notice Provides functions for requesting flash loans to close perps V3 positions.
+ */
+interface IPerpsFlashLoanUtil {
+    /**
+     * @notice Requests a flash loan.
+     * @param _amount The amount of USDC to flash loan.
+     * @param _collateralType The address of the collateral type to use.
+     * @param _marketId The ID of the market for the flash loan.
+     * @param _accountId The account ID to use for the flash loan.
+     */
+    function requestFlashLoan(
+        uint256 _amount,
+        address _collateralType,
+        uint128 _marketId,
+        uint128 _accountId
+    ) external;
+
+    /**
+     * @notice Executes the flash loan operation.
+     * @param asset The address of the asset being flash loaned.
+     * @param amount The amount of the asset being flash loaned.
+     * @param premium The premium to be paid for the flash loan.
+     * @param initiator The address that initiated the flash loan.
+     * @param params Additional parameters for the flash loan operation.
+     * @return success True if the operation was successful, false otherwise.
+     */
+    function executeOperation(
```

Comment:
Can you please tell me which argument is given to ensure that sandwishes are constrained?

Reply:
> Can you please tell me which argument is given to ensure that sandwishes are constrained?

`amountOutMinimum`: https://github.com/Synthetixio/synthetix-v3/blob/b8f01cdaa829f7630c2064ed5717c86d7c3c3dc3/auxiliary/PerpsFlashLoanUtil/contracts/PerpsFlashLoanUtil.sol#L119 

which is derived from uniswap's `IQuoter.quoteExactInput`:
https://github.com/Uniswap/v3-periphery/blob/main/contracts/interfaces/IQuoter.sol

Reply:
ohhh this `amountOutMinimum` is included in `calldata`?

Reply:
> ohhh this `amountOutMinimum` is included in `calldata`?

No, we fetch it from the IQuoter at L119 in `executeOperation`:

```solidity
            uint256 amountOutMinimum = quoter.quoteExactInput(
                abi.encodePacked(collateralType, poolAddress, USDC),
                unwrappedAmount
            );
 ```
 
 And then we use it to do the swap:
 
 ```solidity
             ISwapRouter.ExactInputParams memory swapParams = ISwapRouter.ExactInputParams({
                path: abi.encodePacked(collateralType, poolAddress, USDC),
                recipient: address(this),
                deadline: block.timestamp,
                amountIn: unwrappedAmount,
                amountOutMinimum: amountOutMinimum
            });
            unwrappedAmount = router.exactInput(swapParams);
```

Reply:
> Maybe i'm not clear, but how does the front-end is able to specify a minimum amount of USDC that he expects to receive for selling his wETH/wBTC margin so that he doesn't get rugged with a sandwish

Ah, I see what you mean. Can we mitigate this by quoting the min amount of USDC in requestFlashLoan, and pass it to `params` for execution? The user can be shown the amount of USDC expected to close their position at the time of requesting and expect to get close to that value in a given slippage range.

This way, the user requests a flash loan with a quoted amount in the first block. And then if the price moves down by the following block when the execution is processed, it can compare the initially quoted amount to a new quote and revert if outside the slippage range.

Reply:
Like let's do an example:
- User requests a 1000 usdc flash loan to repay usdc of debt, having 1 ETH margin on hand (3k$)
- Users flash 1000 usdc, 
- Swaps ETH to 2000 usdc, 
- Repays debt
- Repays flash
At no point in this set of transaction is the rate of swap ETH to USDC is specified. Uness we can configure how much ETH is allowed to be spent somewhere, or the rate of ETH to USDC,